nemo_text_processing/text_normalization/zh/taggers/cardinal.py

```diff
@@ -0,0 +1,141 @@
+        3123    -> cardinal{ integer: "三千一百二十三"}
+        3,123   -> cardinal{ integer: "三千一百二十三"}
+        51234   -> cardinal{ integer: "五万一千二百三十四"}
+        51,234  -> cardinal{ integer: "五万一千二百三十四"}
+        0.125   -> cardinal{ integer: "零点一二五"}
+        self.fst:
+        123     -> cardinal{ integer: "一二三" }
+    '''
+
+    def __init__(self, deterministic: bool = True, lm: bool = False):
+        super().__init__(name="cardinal", kind="classify", deterministic=deterministic)
+        # base cardinal from digit to chinese
+        graph_digit = pynini.string_file(get_abs_path("data/number/digit.tsv"))
+        graph_zero = pynini.string_file(get_abs_path("data/number/zero.tsv"))
+        graph_teen = pynini.string_file(get_abs_path("data/number/digit_teen.tsv"))
+        graph_sign = pynini.string_file(get_abs_path("data/number/sign.tsv"))
+
+        graph_digit_with_zero = graph_digit | graph_zero  # from 0(read out) to 9
+        graph_no_zero = pynini.cross("0", "")  # cardinalzero,in some case we don't read it out
+        graph_digit_no_zero = graph_digit | graph_no_zero  # from 0(not read out) to 9
+        insert_zero = pynutil.insert('零')
+        delete_punct = pynutil.delete(",")  # to deal with 5,000,cardinalin english form
+
+        # 15 in 215
+        graph_ten_u = (graph_digit + pynutil.insert(UNIT_1e01) + graph_digit_no_zero) | (graph_zero + graph_digit)
+        # 15 only
+        graph_ten = (graph_teen + pynutil.insert(UNIT_1e01) + graph_digit_no_zero) | (graph_zero + graph_digit)
+        # 215 or 200
+        graph_hundred = (graph_digit + pynutil.insert(UNIT_1e02) + graph_ten_u) | (
+            graph_digit + pynutil.insert(UNIT_1e02) + graph_no_zero ** 2
+        )
+        # 3000 or 3002 or 3012 or 3123
+        graph_thousand = (
+            (graph_digit + pynutil.insert(UNIT_1e03) + graph_hundred)
+            | (
+                graph_digit
+                + pynutil.insert(UNIT_1e03)
+                + graph_zero
+                + graph_digit
+                + pynutil.insert(UNIT_1e01)
+                + graph_digit_no_zero
+            )
+            | (graph_digit + pynutil.insert(UNIT_1e03) + graph_zero + graph_no_zero + graph_digit)
+            | (graph_digit + pynutil.insert(UNIT_1e03) + graph_no_zero ** 3)
+        )
+        # just 3,000 or 3,002 or 3,012 or 3,123
+        graph_thousand_sign = (
+            (graph_digit + pynutil.insert(UNIT_1e03) + delete_punct + graph_hundred)
+            | (
+                graph_digit
+                + pynutil.insert(UNIT_1e03)
+                + delete_punct
+                + graph_zero
+                + graph_digit
+                + pynutil.insert(UNIT_1e01)
+                + graph_digit_no_zero
+            )
+            | (graph_digit + pynutil.insert(UNIT_1e03) + delete_punct + graph_zero + graph_no_zero + graph_digit)
+            | (graph_digit + pynutil.insert(UNIT_1e03) + delete_punct + graph_no_zero ** 3)
+        )
```

Comment:
```
|  closure(...)
 |      closure(self, lower)
 |      closure(self, lower, upper)
 |      
 |      Computes concatenative closure.
 |      
 |      This operation destructively converts the FST to its concatenative closure.
 |      If A transduces string x to y with weight w, then the zero-argument form
 |      `A.closure()` mutates A so that it transduces between empty strings with
 |      weight 1, transduces string x to y with weight w, transduces xx to yy with
 |      weight w \otimes w, string xxx to yyy with weight w \otimes w \otimes w
 |      (and so on).
 |      
 |      When called with two optional positive integer arguments, these act as
 |      lower and upper bounds, respectively, for the number of cycles through the
 |      original FST that the mutated FST permits. Therefore, `A.closure(0, 1)`
 |      mutates A so that it permits 0 or 1 cycles; i.e., the mutated A transduces
 |      between empty strings or transduces x to y.
 |      
 |      When called with one optional positive integer argument, this argument
 |      acts as the lower bound, with the upper bound implicitly set to infinity.
 |      Therefore, `A.closure(1)` performs a mutation roughly equivalent to
 |      `A.closure()` except that the former does not transduce between empty
 |      strings.
 |      
 |      The following are the equivalents for the closure-style syntax used in
 |      Perl-style regular expressions:
 |      
 |      Regexp:             This method:            Copy shortcuts:
 |      
 |      /x?/                x.closure(0, 1)         x.ques
 |      /x*/                x.closure()             x.star
 |      /x+/                x.closure(1)            x.plus
 |      /x{N}/              x.closure(N, N)         x ** N
 |      /x{M,N}/            x.closure(M, N)
 |      /x{N,}/             x.closure(N)
 |      /x{,N}/             x.closure(0, N)
```

Please try to use `x.ques` to avoid the redundant code.